aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp

```diff
@@ -4390,6 +4390,168 @@ void qmul_tensor_cpu_kernel(
+      }
+      if constexpr (std::is_same<T, float>::value) {
+        *(out_ptr + size_com + d) = out_val_f;
+      } else if constexpr (std::is_same<T, at::BFloat16>::value) {
+        *(out_ptr + size_com + d) = at::BFloat16(out_val_f);
+      } else if constexpr (std::is_same<T, at::Half>::value) {
+        *(out_ptr + size_com + d) = at::Half(out_val_f);
+      } else { //  T == uint8, requantization needed
+        out_val_f = std::round(out_val_f * inv_output_scale);
+        int32_t out_val_i32 = (int32_t)out_val_f + output_zero_point;
+        out_val_i32 = std::min(255, std::max(0, out_val_i32));
+        *(out_ptr + size_com + d) = static_cast<uint8_t>(out_val_i32);
+      }
+    }
+  }
+#else
+  at::parallel_for(0, size, 1, [&](int64_t start, int64_t end) {
+    for (const auto d : c10::irange(start, end)) {
+      uint8_t x_data = *(x_ptr + d);
+      uint8_t y_data = *(y_ptr + d);
+      int32_t x_val = static_cast<int32_t>(x_data) - x_zero_point;
+      int32_t y_val = static_cast<int32_t>(y_data) - y_zero_point;
+      float x_val_f = static_cast<float>(x_val) * x_scale;
+      float y_val_f = static_cast<float>(y_val) * y_scale;
+      float out_val_f = x_val_f + y_val_f;
+      if constexpr (ReLUFused) {
+        out_val_f = std::max(out_val_f, 0.f);
+      }
+      if constexpr (std::is_same<T, float>::value) {
+        *(out_ptr + d) = out_val_f;
+      } else if constexpr (std::is_same<T, at::BFloat16>::value) {
+        *(out_ptr + d) = at::BFloat16(out_val_f);
+      } else if constexpr (std::is_same<T, at::Half>::value) {
+        *(out_ptr + d) = at::Half(out_val_f);
+      } else { //  T == uint8, requantization needed
+        out_val_f = std::round(out_val_f * inv_output_scale);
+        int32_t out_val_i32 = (int32_t)out_val_f + output_zero_point;
+        out_val_i32 = std::min(255, std::max(0, out_val_i32));
+        *(out_ptr + d) = static_cast<uint8_t>(out_val_i32);
+      }
+    }
+  });
+#endif
+}
+
+template <bool ReLUFused>
+void qadd_tensor_cpu_kernel(
+    Tensor& out,
+    const Tensor& qx,
+    double qx_scale,
+    int64_t qx_zero_point,
+    const Tensor& qy,
+    double qy_scale,
+    int64_t qy_zero_point,
+    double output_scale,
+    int64_t output_zero_point) {
+  auto qx_ptr = qx.const_data_ptr<uint8_t>();
+  auto qy_ptr = qy.const_data_ptr<uint8_t>();
+  int64_t size = qx.numel();
+  TORCH_CHECK(
+      size == qy.numel() && size == out.numel(),
+      "qadd_cpu: Expect qx, qy and out to have the same number of elements");
+  if (out.scalar_type() == c10::ScalarType::Float) {
+    auto out_ptr = out.data_ptr<float>();
+    _qadd_tensor_cpu_impl<float, ReLUFused>(
+        out_ptr, size, qx_ptr, qx_scale, qx_zero_point, qy_ptr, qy_scale, qy_zero_point, output_scale, output_zero_point
+    );
+  } else if (out.scalar_type() == c10::ScalarType::BFloat16) {
+    auto out_ptr = out.data_ptr<at::BFloat16>();
+    _qadd_tensor_cpu_impl<at::BFloat16, ReLUFused>(
```

Comment:
Can we directly get `at::BFloat16` from `out.scalar_type()`? Then, we can merge these branches.

Reply:
Thanks. I have updated this part.